File: ports/go/internal/demo/components.go
package demo

import ui "github.com/profullstack/hqtui/ports/go"

func (s *state) scrollHandlers(name string, pane *pane) ui.ScrollHandlers {
	return ui.ScrollHandlers{OnFocus: s.action(func() { s.focused[s.screen] = name }), OnScroll: func(d int) {
		if !s.overlay() {
			s.focused[s.screen] = name
			pane.move(d)
		}
	}, OnSelectRow: func(i int) {
		if !s.overlay() {
			pane.selected = clamp(pane.offset+i, 0, pane.total-1)
		}
	}}
}
func processTree() []ui.TreeNode {
	node := func(name, cpu, mem string, children ...ui.TreeNode) ui.TreeNode {
		return ui.TreeNode{Label: name, Values: []ui.TreeValue{{Text: cpu, Width: 6}, {Text: mem, Width: 6}}, Children: children}
	}
	return []ui.TreeNode{node("systemd", "1.3", "0.1", node("bash", "0.1", "0.2"), node("bun", "32.8", "4.2", node("bun:worker", "12.4", "1.8"), node("bun:worker", "8.7", "1.3")), node("node", "18.1", "2.1", node("node:worker", "6.1", "0.8")), node("postgres", "6.7", "1.8"))}
}
func (s *state) components(p *ui.Container) {
	t := p.Theme()
	c, m, n := obj(s.sample["cpu"]), obj(s.sample["memory"]), obj(s.sample["network"])
	row(p, ui.Fr(1), 1, func(r *ui.Container) {
		r.Column(ui.ColumnOptions{Layout: ui.Layout{Gap: 1}}, func(left *ui.Container) {
			left.Panel(ui.PanelOptions{Title: "Buttons & Inputs", Layout: fixed(13)}, func(p *ui.Container) {
				row(p, ui.Cells(1), 1, func(r *ui.Container) {
					for i, label := range []string{"Primary", "Success", "Warning", "Danger"} {
						width := []int{11, 11, 11, 10}[i]
						var action func()
						if i == 0 {
							action = s.action(func() { s.modal = true })
						}
						r.Button(ui.ButtonOptions{Label: label, Width: width, Variant: []ui.ButtonVariant{ui.ButtonPrimary, ui.ButtonSuccess, ui.ButtonWarning, ui.ButtonDanger}[i]}, action, fixed(width))
					}
					r.Spacer(ui.Fill())
				})
				p.Spacer(ui.Cells(1))
				row(p, ui.Cells(1), 2, func(r *ui.Container) {
					options := []string{"Dark", "Dracula", "Nord", "Tokyo Night"}
					r.Select(ui.SelectOptions{Value: options[s.selectIndex%4], Width: 20, Open: s.selectOpen, Options: options, SelectedIndex: s.selectIndex}, s.action(func() { s.selectOpen = !s.selectOpen }), fixed(20))
					r.Checkbox(ui.CheckboxOptions{Label: "Toggle", Checked: s.toggle, Variant: ui.CheckboxToggle}, s.action(func() { s.toggle = !s.toggle }), fixed(12))
					r.Checkbox(ui.CheckboxOptions{Label: "Checkbox", Checked: s.checked}, s.action(func() { s.checked = !s.checked }), fixed(14))
					r.Spacer(ui.Fill())
				})
				p.Spacer(ui.Cells(1))
				p.TextInput(ui.TextInputOptions{Label: "Search", Value: s.input, Placeholder: "type to filter…", Focused: s.editing}, fixed(1))
				p.Spacer(ui.Cells(1))
				p.Meter(ui.MeterOptions{Label: "Slider", Value: s.slider, Heat: ref(false), Color: &t.Primary})
				p.Progress(ui.ProgressOptions{Label: "Progress", Value: 37, Max: ref(120.), ShowCount: true})
			})
			left.Panel(ui.PanelOptions{Title: "Table Widget"}, func(p *ui.Container) {
				rows := []any{}
				for _, r := range [][4]string{{"src", "4.2 KB", "dir", "2m ago"}, {"test", "1.1 KB", "dir", "5m ago"}, {"package.json", "1.2 KB", "file", "10m ago"}, {"README.md", "3.4 KB", "file", "1h ago"}, {"bun.lockb", "12 KB", "file", "1h ago"}} {
					rows = append(rows, object{"name": r[0], "size": r[1], "type": r[2], "modified": r[3]})
				}
				s.dataTable(p, "components.files", rows, []dataColumn{dc("name", "Name", 0, 10, &t.Primary, false), dc("size", "Size", 9, 0, nil, true), dc("type", "Type", 6, 0, nil, false), dc("modified", "Modified", 10, 0, &t.Muted, true)}, true, false, true)
			})
			left.Panel(ui.PanelOptions{Title: "Log Viewer", Layout: fixed(11)}, func(p *ui.Container) {
				data := arr(s.sample["logs"])
				pane := s.pane("components.logs", len(data))
				pane.log = true
				entries := []ui.LogEntry{}
				for _, v := range data {
					d := obj(v)
					entries = append(entries, ui.LogEntry{Time: scalar(d["time"]), Level: scalar(d["level"]), Message: scalar(d["message"]), Meta: "{" + scalar(d["meta"]) + "}"})
				}
				p.Log(ui.LogOptions{Entries: entries, FromEnd: pane.offset, Scrollbar: true}, ui.ScrollHandlers{OnFocus: s.action(func() { s.focused[s.screen] = "components.logs" }), OnScroll: func(d int) {
					if !s.overlay() {
						s.focused[s.screen] = "components.logs"
						pane.offset = clamp(pane.offset-d, 0, len(data)-1)
					}
				}})
			})
		})
		r.Column(ui.ColumnOptions{Layout: ui.Layout{Gap: 1}}, func(right *ui.Container) {
			right.Panel(ui.PanelOptions{Title: "Process Tree", Layout: fixed(13)}, func(p *ui.Container) {
				row(p, ui.Cells(1), 0, func(r *ui.Container) {
					r.StyledText("Name", ui.TextStyle{Fg: &t.Muted, Bold: true})
					r.StyledText("CPU%   MEM%", ui.TextStyle{Fg: &t.Muted, Bold: true, Align: ui.AlignRight})
				})
				pane := s.pane("components.tree", 8)
				p.Tree(ui.TreeOptions{Nodes: processTree(), Selected: &pane.selected, Offset: &pane.offset, FollowSelection: true}, s.scrollHandlers("components.tree", pane))
			})
			right.Panel(ui.PanelOptions{Title: "Sparklines & Gauges", Layout: fixed(12)}, func(p *ui.Container) {
				p.Sparkline(ui.SparklineWidgetOptions{Label: "CPU ", Values: values(c["history"]), Text: percent(num(c["total"])), Color: &t.Success})
				p.Sparkline(ui.SparklineWidgetOptions{Label: "Mem ", Values: values(m["history"]), Text: percent(num(m["used"]) / num(m["total"])), Color: &t.Warning})
				p.Sparkline(ui.SparklineWidgetOptions{Label: "Net ", Values: values(n["downHistory"]), Text: bytes(num(n["downRate"]), 2) + "/s", Color: &t.Primary})
				p.Spacer(ui.Cells(1))
				row(p, ui.Fr(1), 2, func(r *ui.Container) {
					r.Gauge(ui.GaugeOptions{Value: num(c["total"]), Label: percent(num(c["total"]))})
					r.Donut(ui.DonutOptions{Segments: []ui.DonutSegment{{Value: num(m["used"]), Color: &t.Primary, Label: "Used"}, {Value: num(m["available"]), Color: &t.Warning, Label: "Free"}}})
				})
			})
			right.Panel(ui.PanelOptions{Title: "Lists & Badges"}, func(p *ui.Container) {
				row(p, ui.Cells(1), 1, func(r *ui.Container) {
					r.Badge(ui.BadgeOptions{Text: "active", Color: &t.Success}, fixed(10))
					r.Badge(ui.BadgeOptions{Text: "idle", Color: &t.Warning, Variant: ui.BadgeSubtle}, fixed(8))
					r.Badge(ui.BadgeOptions{Text: "failed", Color: &t.Danger, Variant: ui.BadgeOutline}, fixed(10))
					r.Spacer(ui.Fill())
				})
				p.Spacer(ui.Cells(1))
				pane := s.pane("components.list", 4)
				p.List(ui.ListOptions{Items: []ui.ListItem{{Label: "apps/demo", Color: &t.Primary}, {Label: "packages/hqtui"}, {Label: "apps/web"}, {Label: "docs"}}, Selected: &pane.selected, Offset: &pane.offset, FollowSelection: true, Bullet: "▸", Scrollbar: true}, s.scrollHandlers("components.list", pane))
			})
		})
	})
}

File: ports/rust/demo/src/components.rs
//! Native translation of apps/demo/src/screens/components.ts.
use crate::{
    dashboard::*,
    model::*,
    telemetry::{col, table},
};
use hqtui::{graphics::BarStyle, prelude::*, unicode::Align};
use serde_json::json;

fn node(label: &str, a: &str, b: &str, children: Vec<TreeNode>) -> TreeNode {
    TreeNode {
        label: label.into(),
        values: vec![
            TreeValue {
                text: a.into(),
                width: 6,
                ..Default::default()
            },
            TreeValue {
                text: b.into(),
                width: 6,
                ..Default::default()
            },
        ],
        children,
        expanded: Some(true),
        ..Default::default()
    }
}
pub fn render<'a>(ui: &mut Container<'a>, s: &'a State) {
    let gap = s.panel_gap();
    ui.row(Row::new().gap(gap), move |r| {
        r.column(Column::new().gap(gap).bordered(true), move |left| {
            left.panel(Panel::new().title("Buttons & Inputs").size(13), move |p| {
                p.row(Row::new().size(1).gap(1), |r| {
                    for (label, width, variant) in [
                        ("Primary", 11, ButtonVariant::Primary),
                        ("Success", 11, ButtonVariant::Success),
                        ("Warning", 11, ButtonVariant::Warning),
                        ("Danger", 10, ButtonVariant::Danger),
                    ] {
                        r.sized(width, move |c| {
                            c.button(
                                ButtonOptions {
                                    label: label.into(),
                                    width: Some(width),
                                    variant,
                                    ..Default::default()
                                },
                                &format!("button:{label}"),
                            );
                        });
                    }
                    r.spacer("fill");
                });
                p.spacer(1);
                p.row(Row::new().size(1).gap(2), move |r| {
                    r.sized(20, move |c| {
                        let names = ["Dark", "Dracula", "Nord", "Tokyo Night"];
                        c.select(
                            SelectOptions {
                                value: names[s.select_index.min(3)].into(),
                                width: Some(20),
                                open: s.select_open,
                                options: names.iter().map(|v| v.to_string()).collect(),
                                selected_index: Some(s.select_index),
                                ..Default::default()
                            },
                            "select",
                        );
                    });
                    r.sized(12, move |c| {
                        c.checkbox(
                            CheckboxOptions::new("Toggle", s.toggle)
                                .variant(CheckboxVariant::Toggle),
                            "toggle",
                        );
                    });
                    r.sized(14, move |c| {
                        c.checkbox(CheckboxOptions::new("Checkbox", s.checked), "check");
                    });
                    r.spacer("fill");
                });
                p.spacer(1);
                p.text_input(
                    TextInputOptions {
                        label: Some("Search".into()),
                        value: s.input.clone(),
                        placeholder: Some("type to filter…".into()),
                        focused: s.editing,
                        ..Default::default()
                    },
                    "input",
                );
                p.spacer(1);
                p.meter(MeterOptions {
                    label: Some("Slider".into()),
                    value: 0.7,
                    style: Some(BarStyle::Smooth),
                    heat: Some(false),
                    color: Some(p.theme().primary),
                    ..Default::default()
                });
                p.progress(ProgressOptions {
                    label: Some("Progress".into()),
                    value: 37.,
                    max: Some(120.),
                    show_count: true,
                    ..Default::default()
                });
            });
            left.panel(Panel::new().title("Table Widget"), move |p| {
                let data = json!([
                    {"name":"src","size":"4.2 KB","type":"dir","modified":"2m ago"},
                    {"name":"test","size":"1.1 KB","type":"dir","modified":"5m ago"},
                    {"name":"package.json","size":"1.2 KB","type":"file","modified":"10m ago"},
                    {"name":"README.md","size":"3.4 KB","type":"file","modified":"1h ago"},
                    {"name":"bun.lockb","size":"12 KB","type":"file","modified":"1h ago"}
                ]);
                let t = p.theme();
                let cols = vec![
                    col("name", "Name", 0, 10, Some(t.primary), false),
                    col("size", "Size", 9, 0, None, true),
                    col("type", "Type", 6, 0, None, false),
                    col("modified", "Modified", 10, 0, Some(t.muted), true),
                ];
                table(
                    p,
                    s,
                    "components.files",
                    array(&data).to_vec(),
                    cols,
                    true,
                    true,
                );
            });
            left.panel(Panel::new().title("Log Viewer").size(11), move |p| {
                let logs = array(&s.sample["logs"]);
                let mut panes = s.panes.borrow_mut();
                let pane = panes.entry("components.logs".into()).or_default();
                pane.total = logs.len();
                pane.log = true;
                p.log(
                    LogOptions {
                        entries: logs
                            .iter()
                            .map(|l| LogEntry {
                                time: Some(text(&l["time"])),
                                level: Some(text(&l["level"])),
                                message: text(&l["message"]),
                                meta: Some(format!("{{{}}}", text(&l["meta"]))),
                                ..Default::default()
                            })
                            .collect(),
                        from_end: Some(pane.offset),
                        scrollbar: true,
                        ..Default::default()
                    },
                    "components.logs",
                );
            });
        });
        r.column(Column::new().gap(gap).bordered(true), move |right| {
            right.panel(Panel::new().title("Process Tree").size(13), move |p| {
                let muted = p.theme().muted;
                p.row(Row::new().size(1), move |r| {
                    r.styled_text("Name", TextStyle::new().fg(muted).bold());
                    r.styled_text(
                        "CPU%   MEM%",
                        TextStyle::new().fg(muted).bold().align(Align::Right),
                    );
                });
                let mut panes = s.panes.borrow_mut();
                let pane = panes.entry("components.tree".into()).or_default();
                pane.total = 8;
                let roots = vec![node(
                    "systemd",
                    "1.3",
                    "0.1",
                    vec![
                        node("bash", "0.1", "0.2", vec![]),
                        node(
                            "bun",
                            "32.8",
                            "4.2",
                            vec![
                                node("bun:worker", "12.4", "1.8", vec![]),
                                node("bun:worker", "8.7", "1.3", vec![]),
                            ],
                        ),
                        node(
                            "node",
                            "18.1",
                            "2.1",
                            vec![node("node:worker", "6.1", "0.8", vec![])],
                        ),
                        node("postgres", "6.7", "1.8", vec![]),
                    ],
                )];
                pane.offset = resolve_offset(
                    Some(pane.offset),
                    Some(pane.selected),
                    p.height().saturating_sub(1),
                    8,
                    true,
                );
                p.tree(
                    TreeOptions {
                        nodes: roots,
                        selected: Some(pane.selected),
                        offset: Some(pane.offset),
                        follow_selection: true,
                        ..Default::default()
                    },
                    "components.tree",
                );
            });
            right.panel(
                Panel::new().title("Sparklines & Gauges").size(12),
                move |p| {
                    let t = p.theme().clone();
                    let d = &s.sample;
                    for (label, v, readout, c) in [
                        (
                            "CPU ",
                            &d["cpu"]["history"],
                            percent(n(&d["cpu"]["total"])),
                            t.success,
                        ),
                        (
                            "Mem ",
                            &d["memory"]["history"],
                            percent(n(&d["memory"]["used"]) / n(&d["memory"]["total"]).max(1.)),
                            t.warning,
                        ),
                        (
                            "Net ",
                            &d["network"]["downHistory"],
                            format!("{}/s", bytes(n(&d["network"]["downRate"]), 2)),
                            t.primary,
                        ),
                    ] {
                        p.sparkline(SparklineWidgetOptions {
                            label: Some(label.into()),
                            values: values(v),
                            text: Some(readout),
                            color: Some(c),
                            ..Default::default()
                        });
                    }
                    p.spacer(1);
                    p.row(Row::new().gap(2), move |r| {
                        r.gauge(GaugeOptions {
                            value: n(&d["cpu"]["total"]),
                            label: Some(percent(n(&d["cpu"]["total"]))),
                            ..Default::default()
                        });
                        r.donut(DonutOptions {
                            segments: vec![
                                DonutSegment {
                                    value: n(&d["memory"]["used"]),
                                    color: Some(t.primary),
                                    label: Some("Used".into()),
                                },
                                DonutSegment {
                                    value: n(&d["memory"]["available"]),
                                    color: Some(t.warning),
                                    label: Some("Free".into()),
                                },
                            ],
                            ..Default::default()
                        });
                    });
                },
            );
            right.panel(Panel::new().title("Lists & Badges"), move |p| {
                let t = p.theme().clone();
                p.row(Row::new().size(1).gap(1), move |r| {
                    for (label, color, width, variant) in [
                        ("active", t.success, 10, BadgeVariant::Filled),
                        ("idle", t.warning, 8, BadgeVariant::Subtle),
                        ("failed", t.danger, 10, BadgeVariant::Outline),
                    ] {
                        r.sized(width, move |c| {
                            c.badge(BadgeOptions::new(label).color(color).variant(variant));
                        });
                    }
                    r.spacer("fill");
                });
                p.spacer(1);
                let mut panes = s.panes.borrow_mut();
                let pane = panes.entry("components.list".into()).or_default();
                pane.total = 4;
                pane.offset = resolve_offset(
                    Some(pane.offset),
                    Some(pane.selected),
                    p.height().saturating_sub(2),
                    4,
                    true,
                );
                p.list(
                    ListOptions {
                        items: vec![
                            ListItem {
                                label: "apps/demo".into(),
                                color: Some(t.primary),
                                ..Default::default()
                            },
                            ListItem::new("packages/hqtui"),
                            ListItem::new("apps/web"),
                            ListItem::new("docs"),
                        ],
                        selected: Some(pane.selected),
                        offset: Some(pane.offset),
                        follow_selection: true,
                        bullet: Some("▸".into()),
                        scrollbar: true,
                        ..Default::default()
                    },
                    "components.list",
                );
            });
        });
    });
}
